drivers/docker/docker_client.go

```diff
@@ -262,47 +246,55 @@ func (d *dockerWrap) StartContainerWithContext(id string, hostConfig *docker.Hos
 }
 
 func (d *dockerWrap) CreateContainer(opts docker.CreateContainerOptions) (c *docker.Container, err error) {
-	err = d.retry(func() error {
-		c, err = d.dockerLongTimeout.CreateContainer(opts)
+	err = d.retry(opts.Context, func() error {
+		c, err = d.dockerNoTimeout.CreateContainer(opts)
 		return err
 	})
 	return c, err
 }
 
-func (d *dockerWrap) RemoveContainer(opts docker.RemoveContainerOptions) (err error) {
-	err = d.retry(func() error {
-		err = d.docker.RemoveContainer(opts)
+func (d *dockerWrap) PullImage(opts docker.PullImageOptions, auth docker.AuthConfiguration) (err error) {
+	err = d.retry(opts.Context, func() error {
+		err = d.dockerNoTimeout.PullImage(opts, auth)
 		return err
 	})
-	return filterNoSuchContainer(err)
+	return err
 }
 
-func (d *dockerWrap) PullImage(opts docker.PullImageOptions, auth docker.AuthConfiguration) (err error) {
-	err = d.retry(func() error {
-		err = d.dockerLongTimeout.PullImage(opts, auth)
+func (d *dockerWrap) RemoveContainer(opts docker.RemoveContainerOptions) (err error) {
+	ctx, cancel := context.WithTimeout(context.Background(), retryTimeout)
+	defer cancel()
+	err = d.retry(ctx, func() error {
+		err = d.docker.RemoveContainer(opts)
 		return err
 	})
-	return err
+	return filterNoSuchContainer(err)
 }
 
 func (d *dockerWrap) InspectImage(name string) (i *docker.Image, err error) {
-	err = d.retry(func() error {
+	ctx, cancel := context.WithTimeout(context.Background(), retryTimeout)
+	defer cancel()
+	err = d.retry(ctx, func() error {
 		i, err = d.docker.InspectImage(name)
 		return err
 	})
 	return i, err
 }
 
 func (d *dockerWrap) InspectContainer(id string) (c *docker.Container, err error) {
-	err = d.retry(func() error {
+	ctx, cancel := context.WithTimeout(context.Background(), retryTimeout)
+	defer cancel()
+	err = d.retry(ctx, func() error {
```

Comment:
we're using an underlying docker timeout of 2 minutes so it's bounded pretty tightly to basically a worst case of 11 minutes and 59 seconds. we don't really want/need to retry for up to the task timeout for most operations (just the special few that seem to exceed 2 minutes), the context methods were mostly for the methods that need longer timeouts, though yea it creates quite a jungle of timeouts, imo it's still slightly better than having to specify 2 contexts in each method for `dockerClient`.